### ISSUE TITLE
[BUG] X flag of ASR and LSR

### ISSUE THREAD

**Issue by SteveKwok** [kstenerud/Musashi#27]
Hi Karl,

The document says "The X-bit is set to the last bit shifted out of the operand and is
equal to the C-bit. However, a zero shift count leaves the X-bit
unaffected and the C-bit cleared."

The current implementation of Musashi certainly clears the X flag when shift count is zero (when it is stored in a data register).
```
FLAG_X = FLAG_C = src << (9-shift);
```
It should be like the following
```
FLAG_C = src << (9-shift);
if (shift)
{
    FLAG_X = FLAG_C;
}
```
I'm sorry for creating this issue instead of making a pull request.

Regards,
Steve

**Comment by marhel** [comment]
I believe the current implementation is in fact correct, and that if statement isn't needed.

There's two types of shifts, shifting a data register (Dy), or memory.
For the Register shifts, there's two variants - the shift count can be taken either from another data register (Dx), or an immediate constant specified in three bits of the instruction opcode.

The only type that will ever encounter a shift count of zero are the Register/Register shifts, which already have a if-statement for the zero shift case. It's not needed for the other shifts, as Memory shifts can only shift one bit at a time, and Immediate shifts can only shift 1-8 bits. An immediate constant of zero is treated as a shift count of eight, as per the M68000PRM. 

In other words, you cannot shift by a literal zero, and In EASy68k, the instruction `ASR #0, D0` fails with the error "invalid constant shift count", as expected.

Let's look at ASR, but it is the same for LSR.
* Immediate - `ASR #n, Dy` - The shift count (1-8) is a constant specified in the instruction, shifts Dy
    implemented at https://github.com/kstenerud/Musashi/blob/master/m68k_in.c#L1829-L1883
* Register - `ASR Dx, Dy` - The shift count is in Dx, shifts Dy
    implemented at https://github.com/kstenerud/Musashi/blob/master/m68k_in.c#L1886-L2039
* Memory - `ASR <ea>` - Memory at EA is shifted one bit
    implemented at https://github.com/kstenerud/Musashi/blob/master/m68k_in.c#L2042-L2057

**Comment by kstenerud** [comment]
Yes, one must be careful about correctness according to the manual, and how it was actually implemented in the chip. I did a fair bit of direct comparisons to the real thing, although I don't have access to the hardware anymore. Ideally, we'd want this verified according to real hardware to make sure we're not about to break things :)

**Comment by SteveKwok** [comment]
@marhel Thanks for the detailed reply :)
I've not noticed that there is already an "if" statement before the assignment.

@kstenerud Thanks for the tips :)

 So, we got the conclusion, this is not a bug, the issue should be closed, thanks again :)

**Comment by SteveKwok** [comment]
Not a bug
